File: 프로그래머스/level2/디펜스 게임/solution.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int solution(int n, int k, vector<int> enemy) {
    int answer = 0;
    priority_queue<int> pq;
    
    while (answer < enemy.size()) {
        n -= enemy[answer];
        pq.push(enemy[answer]);
        
        if (n < 0) {
            if (k < 1)
                return answer;
            while (n < 0 && k >= 1) {
                n += pq.top();
                pq.pop();
                k--;
            }
            if (n < 0)
                return answer;
        }
        answer++;
    }
    
    return answer;
}

File: 프로그래머스/level2/디펜스 게임/solution.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(int n, int k, int[] enemy) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        int cur = 0;
        while (cur < enemy.length) {
            n -= enemy[cur];
            pq.add(enemy[cur]);
            
            if (n < 0) {
                if (k < 1)
                    return cur;
                while (n < 0 && k > 0 && pq.size() > 0) {
                    Integer top = pq.poll();
                    n += top;
                    k--;
                }
                if (n < 0)
                    return cur;
            }
            cur++;
        }
        return cur;
    }
}
